tools/osbuild-depsolve-dnf5

```diff
@@ -0,0 +1,500 @@
+        if "baseurl" in desc:
+            conf.baseurl = desc["baseurl"]
+        elif "metalink" in desc:
+            conf.metalink = desc["metalink"]
+        elif "mirrorlist" in desc:
+            conf.mirrorlist = desc["mirrorlist"]
+        else:
+            raise ValueError("missing either `baseurl`, `metalink`, or `mirrorlist` in repo")
+
+        if desc.get("ignoressl", False):
+            conf.sslverify = False
+        if "sslcacert" in desc:
+            conf.sslcacert = desc["sslcacert"]
+        if "sslclientkey" in desc:
+            conf.sslclientkey = desc["sslclientkey"]
+        if "sslclientcert" in desc:
+            conf.sslclientcert = desc["sslclientcert"]
+
+        if "check_gpg" in desc:
+            conf.gpgcheck = desc["check_gpg"]
+        if "check_repogpg" in desc:
+            conf.repo_gpgcheck = desc["check_repogpg"]
+        if "gpgkey" in desc:
+            conf.gpgkey = [desc["gpgkey"]]
+        if "gpgkeys" in desc:
+            # gpgkeys can contain a full key, or it can be a URL
+            # dnf expects urls, so write the key to a temporary location and add the file://
+            # path to conf.gpgkey
+            keydir = os.path.join(self.base.get_config().persistdir, "gpgkeys")
+            if not os.path.exists(keydir):
+                os.makedirs(keydir, mode=0o700, exist_ok=True)
+
+            for key in desc["gpgkeys"]:
+                if key.startswith("-----BEGIN PGP PUBLIC KEY BLOCK-----"):
+                    # Not using with because it needs to be a valid file for the duration. It
+                    # is inside the temporary persistdir so will be cleaned up on exit.
+                    # pylint: disable=consider-using-with
+                    keyfile = tempfile.NamedTemporaryFile(dir=keydir, delete=False)
+                    keyfile.write(key.encode("utf-8"))
+                    conf.gpgkey += (f"file://{keyfile.name}",)
+                    keyfile.close()
+                else:
+                    conf.gpgkey += (key,)
+
+        # In dnf, the default metadata expiration time is 48 hours. However,
+        # some repositories never expire the metadata, and others expire it much
+        # sooner than that. We therefore allow this to be configured. If nothing
+        # is provided we error on the side of checking if we should invalidate
+        # the cache. If cache invalidation is not necessary, the overhead of
+        # checking is in the hundreds of milliseconds. In order to avoid this
+        # overhead accumulating for API calls that consist of several dnf calls,
+        # we set the expiration to a short time period, rather than 0.
+        conf.metadata_expire = desc.get("metadata_expire", "20s")
+
+        # This needs to be set on each repo
+        repo.set_substitutions(self.substitutions)
+
+        # Set the packages to exclude
+        conf.excludepkgs = exclude_pkgs
+
+        return repo
+
+    @staticmethod
+    def _timestamp_to_rfc3339(timestamp):
```

Comment:
(super nitpick) given that this is not specific to DNF I would just move it out of the class into a helper function in the file but given that it's symetric to osbuild-depsolve-dnf it's fine to ignore :)

Reply:
yeah, I'd rather leave it as-is since it matches `osbuild-depsolve-dnf` and is just a simple helper.